packages/test/package.json

```diff
@@ -21,6 +21,9 @@
     ],
     "./package.json": "./package.json"
   },
+  "imports": {
+    "#internal/*": "./lib/*.js"
+  },
```

Comment:
this feature requires node v12.19+ and v14.6+ - do the build scripts handle it for the general case?

Reply:
All node uses commonjs bundle. Imports are only for other tools like webpack

Reply:
See line 12

Reply:
I’m confused; what would other tools use this for?

Marking a thing with “imports” doesn’t make it private; “not putting it in exports” does that.

Reply:
Putting it in imports makes it importable by another module. Right now we pack complains when the exported module tries to import another file

Reply:
hmm, that’s confusing. You’re saying this is how webpack wants you to mark “externals” that are relative files?

Reply:
```js
// good
import { MockResolver } from '@rest-hooks/test';

// bad
import MockResolver from '@rest-hooks/test/MockResolver';
```

I don't want any exports besides the top level. However, I also don't want everything rolled up in one file or webpack won't be able to tree shake. Imports are needed so the export file can import another module inside - without exposing it to top level exports.

Reply:
Right - but that last part isn’t how node works. You can import anything inside the package you want without putting it in exports or imports if it’s a relative path. It’s only if you’re using the package’s own name (the self-reference feature) that “exports” restricts internal imports and you’d need “imports”. Is there a reason the bundle isn’t using relative paths?

Reply:
It is using relative paths. I guess webpack diverges from nodes spec then?

```export { default as MockResolver } from './MockResolver';```

Webpack doesn't like this tho. Maybe you should file an issue with them?

Reply:
```
[webpack-cli] ModuleNotFoundError: Module not found: Error: Can't resolve './MockProvider' in '/home/ntucker/src/anansi/node_modules/@rest-hooks/test/lib'
anansi-example-typescript: Did you mean 'MockProvider.js'?
anansi-example-typescript: BREAKING CHANGE: The request './MockProvider' failed to resolve only because it was resolved as fully specified
anansi-example-typescript: (probably because the origin is a '*.mjs' file or a '*.js' file where the package.json contains '"type": "module"').
anansi-example-typescript: The extension in the request is mandatory for it to be fully specified.**
```


oh wait does this just mean i need to include the .js at the end?

Reply:
Yes, ESM doesn't resolve extensions.

Reply:
What's a good tool to add js for js files? I can't just add it manually as they start off as .ts files

Reply:
If babel is doing the typescript transpilation (as it should be) then you should be able to use https://www.npmjs.com/package/babel-plugin-module-resolver to transpile and add extensions.

Your current approach should work as well.

Reply:
Is there a project already using that package? I was having trouble writing the alias regex

Reply:
I typically only see it in applications; I haven't used it in projects myself. I'd expect using https://github.com/tleunen/babel-plugin-module-resolver/blob/HEAD/DOCS.md#regular-expressions would work with something like `"^(@rest\/hooks\/test.+)(\.[jt]sx?)?$": "\\1.js"`?